File: 2024/day02.hs
import Data.List
import System.IO
import Text.Printf
import Data.Char

isSuperSafe :: [Int] -> Bool
isSuperSafe report = 
    let diffs = zipWith (-) report (tail report)
        signs = map signum diffs
    in all (\n -> n >= 1 && n <= 3) (map abs diffs) && all (==head signs) (tail signs)

part1 :: [[Int]] -> Int
part1 input = length $ filter isSuperSafe input

removeIndex :: Int -> [Int] -> [Int]
removeIndex idx lst = 
    let (a, b) = splitAt idx lst
    in a ++ tail b

isSafe :: [Int] -> Bool
isSafe report = any isSuperSafe $ map (\idx -> removeIndex idx report) [0..length report - 1]

part2 :: [[Int]] -> Int
part2 input = length $ filter isSafe input

main :: IO ()
main = do
    input <- map (map read . words) . lines <$> readFile "./input.txt"
    print input
    printf "part 1: %d\n" (part1 input)
    printf "part 2: %d\n" (part2 input)

File: 2024/day02.cs
namespace AdventOfCode
{
    class Solution
    {
        static bool IsSuperSafe(IEnumerable<int> report)
        {
            var gaps = report.Skip(1).Zip(report, (a, b) => a - b);
            var signs = gaps.Select(Math.Sign);
            return gaps.Select(Math.Abs).All(n => n >= 1 && n <= 3) 
                && signs.Skip(1).All(n => n == signs.First());
        }

        static bool IsSafe(IEnumerable<int> report)
            => report.Select((_, index) => report.Where((_, idx) => index != idx))
                .Any(IsSuperSafe);

        static int Part1(IEnumerable<IEnumerable<int>> input)
            => input.Where(IsSuperSafe).Count();

        static int Part2(IEnumerable<IEnumerable<int>> input)
            => input.Where(IsSafe).Count();

        static IEnumerable<IList<int>> Parse(StreamReader stream)
        {
            string? line;
            while ((line = stream.ReadLine()) != null) 
            {
                yield return line.Split(' ').Select(Int32.Parse).ToList();
            }
        }

        static void Main()
        {
            var input = Parse(new StreamReader("./input.txt")).ToList();
            Console.WriteLine($"part 1: {Part1(input)}");
            Console.WriteLine($"part 2: {Part2(input)}");
        }
    }
}
